Test/Algorithms/DirectSearch/Operators/HypervolumeContribution.cpp

```diff
@@ -184,4 +190,74 @@ BOOST_AUTO_TEST_CASE( Algorithms_HypervolumeContribution3D ) {
 	}
 }
 
+BOOST_AUTO_TEST_CASE( Algorithms_HypervolumeContributionMD_With_3D ) {
+
+	HypervolumeContributionMD hs;
+	const unsigned int numTests = 20;
+	const std::size_t numPoints = 100;
+	
+	RealVector reference(3,1.0);
+	Rng::seed(42);
+	
+	for(unsigned int t = 0; t != numTests; ++t){
+		auto frontLinear = createRandomFront(numPoints,3,1);
+		auto frontConvex = createRandomFront(numPoints,3,2);
+		auto frontConcave = createRandomFront(numPoints,3,0.5);
+		
+		for(std::size_t k = 1; k <= 7; ++k){
+			testContribution(hs,frontLinear,k,reference);
+			testContribution(hs,frontConvex,k,reference);
+			testContribution(hs,frontConcave,k,reference);
+			testContributionNoRef(hs,frontLinear,k);
+			testContributionNoRef(hs,frontConvex,k);
+			testContributionNoRef(hs,frontConcave,k);
+		}
+		
+		//all points
+		testContribution(hs,frontLinear,numPoints,reference);
+		testContribution(hs,frontConvex,numPoints,reference);
+		testContribution(hs,frontConcave,numPoints,reference);
+		
+		//without a reference we can not get moe than numPoints-3 contributions
+		//as three points are required to set up the reference frame
+		testContributionNoRef(hs,frontLinear,numPoints-3);
+		testContributionNoRef(hs,frontConvex,numPoints-3);
+		testContributionNoRef(hs,frontConcave,numPoints-3);
+	}
+}
+
+
+BOOST_AUTO_TEST_CASE( Algorithms_HypervolumeContributionApproximator ) {
+	const unsigned int numTests = 20;
+	const unsigned int numTrials = 100;
+	const std::size_t numPoints = 20;
+	
+	RealVector reference(3,1.0);
+	Rng::seed(42);
+	
+	for(unsigned int t = 0; t != numTests; ++t){
+		auto set = createRandomFront(numPoints,3,2);
+		
+		auto contributionsTrue = contributionsNaive(set, reference);
+		std::vector<double> contributions(set.size());
+		for(std::size_t i = 0; i != contributionsTrue.size(); ++i){
+			contributions[contributionsTrue[i].value]=contributionsTrue[i].key;
+		}
+
+		HypervolumeContributionApproximator algorithm;
+		algorithm.epsilon() = 0.01;
+		algorithm.delta() = 0.1;
+		std::vector<double> approxContributions;
+		for(std::size_t i = 0; i != numTrials; ++i){
+			auto result = algorithm.smallest(set,1,reference)[0].value;
+			approxContributions.push_back(contributions[result]);
+		}
+		std::sort(approxContributions.begin(),approxContributions.end());
+		
+		//check that we do not have too many errors, i.e. contributions with errors larger than 1+epsilon
+		//we make on average 100*errorProbability=10 errors. we give 50% more slack
+		BOOST_CHECK_LT(approxContributions[(1-1.5*algorithm.delta())*numTrials], (1+algorithm.epsilon())*contributionsTrue[0].key);
```

Comment:
Hmmm... that's not a good test. In the worst case the number of failures is a binomial distribution with p=0.1. Hence the variance is np(1-p) = 9, stddev=3, and the slack of 5 is not even two standard deviations.
1. besser express the slack as a multiple of the standard deviation,
2. better use three standard deviations to be on the safe side.
This shifts the threshold from 15 to 19, which is still a meaningful test.